tikv/dbreader/db_reader.go

```diff
@@ -38,6 +42,72 @@ type DBReader struct {
 	safePoint uint64
 }
 
+type RecordWithMeta struct {
+	StartTs  uint64
+	CommitTs uint64
+	Value    []byte
+}
+
+func (r *DBReader) GetWithMeta(key []byte, startTs uint64) (*RecordWithMeta, error) {
+	item, err := r.txn.Get(key)
+	if err != nil && err != badger.ErrKeyNotFound {
+		return nil, errors.Trace(err)
+	}
+	if err == badger.ErrKeyNotFound {
+		return nil, nil
+	}
+	dbUsrMeta := mvcc.DBUserMeta(item.UserMeta())
+	if dbUsrMeta.CommitTS() <= startTs {
+		val, err := item.Value()
+		if err != nil {
+			return nil, err
+		}
+		res := &RecordWithMeta{
+			StartTs:  dbUsrMeta.StartTS(),
+			CommitTs: dbUsrMeta.CommitTS(),
+			Value:    val,
+		}
+		return res, nil
+	}
+	return r.GetOldWithMeta(key, startTs)
+}
+
+func (r *DBReader) GetOldWithMeta(key []byte, startTs uint64) (*RecordWithMeta, error) {
```

Comment:
This method better not export.

Reply:
Or we scan all the old versions for the key instead of lookup many times.

Reply:
Changed to look for the current version, and then scan all historical versions